src/main/kotlin/org/pkl/lsp/documentation/toMarkdown.kt

```diff
@@ -0,0 +1,219 @@
+          } else {
+            val parent = this@doRenderMarkdown.parent
+            if (parent != null && parent is PklMethod) {
+              val type = parent.body.computeExprType(project.pklBaseModule, mapOf(), context)
+              type.render()
+            } else "unknown"
+          }
+        append(": ")
+        append(returnTypeStr)
+      }
+    is PklParameterList -> {
+      elements.joinToString(", ", prefix = "(", postfix = ")") {
+        it.doRenderMarkdown(originalNode, context)
+      }
+    }
+    is PklTypeParameterList -> {
+      typeParameters.joinToString(", ", prefix = "<", postfix = ">") {
+        it.doRenderMarkdown(originalNode, context)
+      }
+    }
+    is PklTypeAnnotation -> ": ${type!!.doRenderMarkdown(originalNode, context)}"
+    is PklTypedIdentifier ->
+      renderTypeAnnotation(identifier?.text, typeAnnotation?.type, this, originalNode, context)!!
+    is PklTypeParameter -> {
+      val vari = variance?.name?.lowercase()?.let { "$it " } ?: ""
+      "$vari$name"
+    }
+    is PklClass ->
+      buildString {
+        append(modifiers.render())
+        append("class ")
+        append(identifier?.text ?: "<class>")
+        typeParameterList?.let { append(it.doRenderMarkdown(originalNode, context)) }
+        supertype?.let {
+          append(" extends ")
+          append(it.doRenderMarkdown(originalNode, context))
+        }
+      }
+    is PklModule -> header?.doRenderMarkdown(originalNode, context) ?: "module $moduleName"
+    is PklModuleHeader ->
+      buildString {
+        append(modifiers.render())
+        append("module ")
+        // can never be null
+        moduleClause?.let { append(it.doRenderMarkdown(originalNode, context)) }
+          ?: append("<module>")
+        moduleExtendsAmendsClause?.let {
+          append(if (it.isAmend) " amends " else " extends ")
+          append(it.moduleUri!!.stringConstant.text)
+        }
+      }
+    is PklModuleClause ->
+      buildString { append(moduleName ?: enclosingModule?.moduleName ?: "<module>") }
+    is PklImport ->
+      buildString {
+        if (isGlob) {
+          append("import* ")
+        } else {
+          append("import ")
+        }
+        moduleUri?.stringConstant?.escapedText()?.let { append("\"$it\"") }
+        val definitionType =
+          resolve(context).computeResolvedImportType(project.pklBaseModule, mapOf(), false, context)
+        append(": ")
+        definitionType.render(this, DefaultTypeNameRenderer)
+      }
+    is PklTypeAlias ->
+      buildString {
+        append(modifiers.render())
+        append("typealias ")
+        append(identifier?.text)
+        typeParameterList?.let { append(it.doRenderMarkdown(originalNode, context)) }
+      }
+    is PklType -> render()
+    else -> text
+  }
+
+// render modifiers
+private fun List<Terminal>?.render(): String {
+  return this?.let { if (isEmpty()) "" else joinToString(" ", postfix = " ") { it.text } } ?: ""
```

Comment:
```suggestion
  return this?.let { joinToString(" ", postfix = " ") { it.text } } ?: ""
```

joining `listOf()` is `""`